roles/ci_dcn_site/tasks/main.yml

```diff
@@ -0,0 +1,252 @@
+      | first }}
+
+- name: Calculate storage network CIDR
+  ansible.builtin.set_fact:
+    _storage_network_range: >-
+      {{ selected_storage_network.ip_v4 }}/{{ selected_storage_network.prefix_length_v4 }}
+
+- name: Calculate storagemgmt network CIDR
+  ansible.builtin.set_fact:
+    _storage_mgmt_network_range: >-
+      {{ selected_storagemgmt_network.ip_v4 }}/{{ selected_storagemgmt_network.prefix_length_v4 }}
+
+- name: Render the pre-ceph NodeSet values.yaml
+  vars:
+    _edpm_instance_dict: "{{ cifmw_networking_env_definition.instances }}"
+    _edpm_network_dict: "{{ cifmw_networking_env_definition.networks }}"
+    _ssh_authorizedkeys: "{{ lookup('file', '~/.ssh/id_cifw.pub', rstrip=False) }}"
+    _ssh_private_key: "{{ lookup('file', '~/.ssh/id_cifw', rstrip=False) }}"
+    _ssh_public_key: "{{ lookup('file', '~/.ssh/id_cifw.pub', rstrip=False) }}"
+    _migration_priv_key: "{{ lookup('file', '~/ci-framework-data/artifacts/nova_migration_key', rstrip=False) }}"
+    _migration_pub_key: "{{ lookup('file', '~/ci-framework-data/artifacts/nova_migration_key.pub', rstrip=False) }}"
+  ansible.builtin.template:
+    backup: true
+    src: "{{ _jinja_template_src }}/edpm-pre-ceph/nodeset/values.yaml.j2"
+    dest: "{{ _arch_deployment_path }}/edpm-pre-ceph/nodeset/values.yaml"
+
+- name: Render the pre-ceph DataPlaneDeployment values.yaml
+  ansible.builtin.template:
+    backup: true
+    src: "{{ _jinja_template_src }}/edpm-pre-ceph/deployment/values.yaml.j2"
+    dest: "{{ _arch_deployment_path }}/edpm-pre-ceph/deployment/values.yaml"
+
+- name: Kustomize pre-ceph NodeSet
+  ansible.builtin.set_fact:
+    nodeset_cr: >-
+      {{ lookup('kubernetes.core.kustomize',
+      dir=_arch_deployment_path + '/edpm-pre-ceph/nodeset') }}
+
+- name: Save the pre-ceph NodeSet CR
+  ansible.builtin.copy:
+    dest: "{{ _arch_deployment_path }}/dataplane-nodeset-pre-ceph_{{ _az }}.yaml"
+    content: "{{ nodeset_cr }}"
+    backup: true
+
+- name: Kustomize pre-ceph DataPlaneDeployment
+  ansible.builtin.set_fact:
+    deployment_cr: >-
+      {{ lookup('kubernetes.core.kustomize',
+      dir=_arch_deployment_path + '/edpm-pre-ceph/deployment') }}
+
+- name: Save the pre-ceph DataPlaneDeployment CR
+  ansible.builtin.copy:
+    dest: "{{ _arch_deployment_path }}/dataplane-deployment-pre-ceph_{{ _az }}.yaml"
+    content: "{{ deployment_cr }}"
+    backup: true
+
+- name: Apply pre-ceph NodeSet CR
+  kubernetes.core.k8s:
+    api_key: "{{ _auth_results.openshift_auth.api_key }}"
+    state: present
+    apply: true
+    src: "{{ _arch_deployment_path }}/dataplane-nodeset-pre-ceph_{{ _az }}.yaml"
+
+- name: Apply pre-ceph DataPlaneDeployment
+  kubernetes.core.k8s:
+    api_key: "{{ _auth_results.openshift_auth.api_key }}"
+    state: present
+    apply: true
+    src: "{{ _arch_deployment_path }}/dataplane-deployment-pre-ceph_{{ _az }}.yaml"
+    wait: true
+    wait_condition:
+      type: Ready
+      status: "True"
+    wait_timeout: 2400
+
```

Comment:
Do we need to refresh the facts here?

https://github.com/openstack-k8s-operators/ci-framework/pull/2458/files#diff-708f8c2df42f7ddde05764ebc7ba803b4fe13ac61c5b3902aff48b7ebde1f4e7R122

Reply:
I don’t think it’s necessary to refresh the facts here. It might be more reasonable to execute the task like this: 
```yaml
- name: Ensure Ceph bootstrap host can ping itself
  ansible.builtin.command:
    cmd: |
      ping -c2 "{{ (_edpm_hosts | dict2items | first).key }}"
  register: _cmd_result
  retries: 5
  delay: 10
  until: _cmd_result.rc == 0
  delegate_to: "{{ (_edpm_hosts | dict2items | first).key }}"
  ```

Reply:
OK, I've added that task.  Let's see if it addresses the problem I ran into.